src/linear_algebra/matrix/fpow.lean

```diff
@@ -0,0 +1,397 @@
+      rw [←neg_add, ←int.coe_nat_add, fpow_neg_coe_nat, fpow_neg_coe_nat,
+          nonsing_inv_apply_not_is_unit, nonsing_inv_apply_not_is_unit, zero_mul],
+      { rwa [det_pow, is_unit_pos_pow_iff],
+        simpa using hm' },
+      { rwa [det_pow, is_unit_pos_pow_iff],
+        rw ←zero_add 0,
+        exact add_lt_add hm' hn' } },
+    { obtain ⟨k, rfl⟩ := eq_coe_of_zero_le hm',
+      obtain ⟨l, rfl⟩ := eq_coe_of_zero_le hn',
+      rw [←int.coe_nat_add, fpow_coe_nat, fpow_coe_nat, fpow_coe_nat, pow_add] } }
+end
+
+lemma fpow_add'' {A : M} {m n : ℤ} (h : is_unit A.det ∨ (m ≤ 0 ∧ n ≤ 0 ∨ 0 < m ∧ 0 < n)) :
+  A ^ (m + n) = A ^ m * A ^ n :=
+begin
+  by_cases ha : is_unit A.det,
+  { exact fpow_add ha m n },
+  { casesI is_empty_or_nonempty n' with H H,
+    { ext i j,
+      exact is_empty_elim i },
+    nontriviality K,
+    simp only [false_or, ha] at h,
+    rcases h with (⟨hm', hn'⟩|⟨hm', hn'⟩),
+    { rcases hm'.eq_or_lt with rfl|hk,
+      { rw [zero_add, gpow_zero, one_mul] },
+      rcases hn'.eq_or_lt with rfl|hl,
+      { rw [add_zero, gpow_zero, mul_one] },
+      obtain ⟨k, rfl⟩ := exists_eq_neg_of_nat hk.le,
+      obtain ⟨l, rfl⟩ := exists_eq_neg_of_nat hl.le,
+      simp only [right.neg_neg_iff, coe_nat_pos] at hk hl,
+      rw [←neg_add, ←int.coe_nat_add, fpow_neg_coe_nat, fpow_neg_coe_nat,
+          nonsing_inv_apply_not_is_unit, nonsing_inv_apply_not_is_unit, zero_mul],
+      { rwa [det_pow, is_unit_pos_pow_iff],
+        simpa using hk },
+      { rwa [det_pow, is_unit_pos_pow_iff],
+        rw ←zero_add 0,
+        exact add_lt_add hk hl } },
+    { obtain ⟨k, rfl⟩ := eq_coe_of_zero_le hm'.le,
+      obtain ⟨l, rfl⟩ := eq_coe_of_zero_le hn'.le,
+      rw [←int.coe_nat_add, fpow_coe_nat, fpow_coe_nat, fpow_coe_nat, pow_add] } }
+end
+
+theorem fpow_one_add {A : M} (h : is_unit A.det) (i : ℤ) : A ^ (1 + i) = A * A ^ i :=
+by rw [fpow_add h, gpow_one]
+
+theorem semiconj_by.fpow_right {a x y : M} (hx : is_unit x.det) (hy : is_unit y.det)
+  (h : semiconj_by a x y) :
+  ∀ m : ℤ, semiconj_by a (x^m) (y^m)
+| (n : ℕ) := by simp [h.pow_right n]
+| -[1+n]  := begin
+  by_cases ha : a = 0,
+  { simp only [ha, semiconj_by.zero_left] },
+  have hx' : is_unit (x ^ n.succ).det,
+  { rw det_pow,
+    exact hx.pow n.succ },
+  have hy' : is_unit (y ^ n.succ).det,
+  { rw det_pow,
+    exact hy.pow n.succ },
+  rw [gpow_neg_succ_of_nat, gpow_neg_succ_of_nat, nonsing_inv_apply _ hx', nonsing_inv_apply _ hy',
+      semiconj_by],
+  refine (is_regular_of_is_left_regular_det hy'.is_regular.left).left _,
+  rw [←mul_assoc, ←(h.pow_right n.succ).eq, mul_assoc, mul_eq_mul (x ^ _), mul_smul, mul_adjugate,
+      mul_eq_mul, mul_eq_mul, mul_eq_mul, ←matrix.mul_assoc, mul_smul (y ^ _), mul_adjugate,
+      units.smul_def, units.smul_def, smul_smul, smul_smul, hx'.coe_inv_mul, hy'.coe_inv_mul,
+      one_smul, matrix.mul_one, matrix.one_mul],
+  apply_instance -- TODO: why is this necessary?
```

Comment:
In the expression `mul_smul (y ^ _)` then `R := units K` can't be inferred yet. So also the instances can't be inferred, and you get a typeclass goal left over. To fix it, add enough parameters to make `R` inferrable again.
```suggestion
  rw [←mul_assoc, ←(h.pow_right n.succ).eq, mul_assoc, mul_eq_mul (x ^ _), mul_smul, mul_adjugate,
      mul_eq_mul, mul_eq_mul, mul_eq_mul, ←matrix.mul_assoc, mul_smul (y ^ _) (hy'.unit)⁻¹, mul_adjugate,
      units.smul_def, units.smul_def, smul_smul, smul_smul, hx'.coe_inv_mul, hy'.coe_inv_mul,
      one_smul, matrix.mul_one, matrix.one_mul],
```